FirebaseStorageSwift/Sources/SwiftApiExtension.swift

```diff
@@ -0,0 +1,96 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import FirebaseStorage
+
+/// getResultCallback generates a closure that returns a Result type from a closure that returns an optional type and Error.
+private func getResultCallback<T>(
+  completion: @escaping (Result<T, Error>) -> Void
+) -> (_: T?, _: Error?) -> Void {
+  return { (metadata: T?, error: Error?) -> Void in
+    guard let metadata = metadata else {
+      guard let error = error else {
+        completion(.failure(NSError(domain: "FirebaseStorageSwift",
+                                    code: -1,
+                                    userInfo: ["Storage Result Generator":
+                                      "InternalError - Return type and Error code both nil"])))
+        return
+      }
+      completion(.failure(error))
+      return
+    }
+    completion(.success(metadata))
+  }
+}
+
+public extension StorageReference {
+  func downloadURL(completion: @escaping (Result<URL, Error>) -> Void) {
+    downloadURL(completion: getResultCallback(completion: completion))
+  }
+
+  func getData(maxSize: Int64, completion: @escaping (Result<Data, Error>) -> Void)
+    -> StorageDownloadTask {
+    return getData(maxSize: maxSize, completion: getResultCallback(completion: completion))
+  }
+
+  func getMetadata(completion: @escaping (Result<StorageMetadata, Error>) -> Void) {
+    getMetadata(completion: getResultCallback(completion: completion))
+  }
+
+  func list(withMaxResults maxResults: Int64,
+            pageToken: String,
+            completion: @escaping (Result<StorageListResult, Error>) -> Void) {
+    list(withMaxResults: maxResults,
+         pageToken: pageToken,
+         completion: getResultCallback(completion: completion))
+  }
+
+  func list(withMaxResults maxResults: Int64,
+            completion: @escaping (Result<StorageListResult, Error>) -> Void) {
+    list(withMaxResults: maxResults,
+         completion: getResultCallback(completion: completion))
+  }
+
+  func listAll(completion: @escaping (Result<StorageListResult, Error>) -> Void) {
+    listAll(completion: getResultCallback(completion: completion))
+  }
+
+  func putData(uploadData: Data,
```

Comment:
Labeling this as `uploadData` will force your callers to also label, when the underlying call does not. This should be

```swift
func putData(_ uploadData: Data, ...);
```

Reply:
Good point. A more Swift friendly name probably will be :
```
func put(data: Data, ...)
```

similar to mentioned below `func write(toFile: URL, ...)`

Reply:
Which is more important? - Choose the more Swift-friendly `put` or keep `putData` to be consistent with the existing Swift API?

I'm leaning towards the former so that it is clear that the two `putData` API's are exactly the same except for their completion parameter.

Reply:
Since you're trying to add an overload, I'd steer for more consistency with the existing API than anything else, though @ryanwilson and the API council's opinion probably matters more than mine.

Reply:
+1 for consistency, and I recall explicitly discussing something similar to `put(_ data: Data)` vs `putData(_ data: Data)` call with @wilhuff during the Firestore review process, and I think `putData(_ data: Data)` is cleaner and more explicit than a simple `put` call.

FWIW @maksymmalyhin I initially argued the same thing (`put` without `Data`) when we first started discussing it 😄

Reply:
Done